core/src/services/zookeeper/backend.rs

```diff
@@ -0,0 +1,305 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::collections::HashMap;
+use zookeeper_client as zk;
+
+use crate::raw::adapters::kv;
+use crate::Scheme;
+use async_trait::async_trait;
+use tokio::sync::OnceCell;
+
+use crate::Builder;
+use crate::Error;
+use crate::ErrorKind;
+use crate::*;
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::io;
+
+use futures::future::{BoxFuture, FutureExt};
+use substring::Substring;
+
+use log::warn;
+
+const DEFAULT_ZOOKEEPER_ENDPOINT: &str = "127.0.0.1:2181";
+/// The scheme for zookeeper authentication
+/// currently we do not support sasl authentication
+const ZOOKEEPER_AUTH_SCHEME: &str = "digest";
+
+/// Zookeeper backend builder
+#[derive(Clone, Default)]
+pub struct ZookeeperBuilder {
+    /// network address of the Zookeeper service
+    /// Default: 127.0.0.1:2181
+    endpoint: Option<String>,
+    /// the user to connect to zookeeper service, default None
+    user: Option<String>,
+    /// the path to the password file of the user to connect to zookeeper service, default None
+    digest_path: Option<String>,
+}
+
+impl ZookeeperBuilder {
+    /// Set the network addresses of zookeeper service
+    pub fn endpoint(&mut self, endpoint: &str) -> &mut Self {
+        if !endpoint.is_empty() {
+            self.endpoint = Some(endpoint.to_string());
+        }
+        self
+    }
+
+    /// Set the username of zookeeper service
+    pub fn user(&mut self, user: &str) -> &mut Self {
+        if !user.is_empty() {
+            self.user = Some(user.to_string());
+        }
+        self
+    }
+
+    /// Specify the auth digest path of zookeeper service
+    pub fn digest_path(&mut self, digest_path: &str) -> &mut Self {
```

Comment:
Is it normal to use `digest_path` to connect zookeeper? How about passing `password` directly?